src/CaptureFile/CaptureFile.cpp

```diff
@@ -0,0 +1,202 @@
+  uint32_t version;
+  uint64_t capture_section_offset;
+  uint64_t section_list_offset;
+};
+
+class CaptureFileImpl : public CaptureFile {
+ public:
+  explicit CaptureFileImpl(std::filesystem::path file_path) : file_path_{std::move(file_path)} {}
+  ~CaptureFileImpl() override = default;
+
+  ErrorMessageOr<void> Initialize();
+
+  ErrorMessageOr<uint16_t> AddSection(uint64_t section_type, uint64_t section_size) override;
+
+  ErrorMessageOr<void> WriteToSection(uint64_t section_number, uint64_t offset_in_section,
+                                      const void* data, size_t size) override;
+
+  [[nodiscard]] const std::vector<CaptureFileSection>& GetSectionList() const override;
+  ErrorMessageOr<void> ReadFromSection(uint64_t section_number, uint64_t offset_in_section,
+                                       void* data, size_t size) override;
+
+ private:
+  // Parse header and validate version/file-format
+  ErrorMessageOr<void> ReadHeaderAndSectionList();
+
+  std::filesystem::path file_path_;
+  orbit_base::unique_fd fd_;
+  CaptureFileHeader header_{};
+  std::vector<CaptureFileSection> section_list_;
+};
+
+ErrorMessageOr<void> CaptureFileImpl::Initialize() {
+  auto fd_or_error = orbit_base::OpenExistingFileForReadWrite(file_path_);
+  if (fd_or_error.has_error()) {
+    return fd_or_error.error();
+  }
+
+  fd_ = std::move(fd_or_error.value());
+
+  OUTCOME_TRY(ReadHeaderAndSectionList());
+  return outcome::success();
+}
+
+ErrorMessageOr<void> CaptureFileImpl::ReadHeaderAndSectionList() {
+  OUTCOME_TRY(header, orbit_base::ReadFullyAtOffset<CaptureFileHeader>(fd_, 0));
+  header_ = header;
+
+  if (std::memcmp(header_.signature.data(), kFileSignature, header_.signature.size()) != 0) {
+    return ErrorMessage{"Invalid file signature"};
+  }
+
+  if (header_.version != kFileVersion) {
+    return ErrorMessage{
+        absl::StrFormat("Incompatible version %d, expected %d", header_.version, kFileVersion)};
+  }
+
+  if (header_.section_list_offset == 0) {
+    return outcome::success();
+  }
+
+  OUTCOME_TRY(number_of_sections,
+              orbit_base::ReadFullyAtOffset<uint16_t>(fd_, header_.section_list_offset));
+  std::vector<CaptureFileSection> section_list{number_of_sections};
+  OUTCOME_TRY(bytes_read,
+              orbit_base::ReadFullyAtOffset(fd_, section_list.data(),
+                                            number_of_sections * sizeof(CaptureFileSection),
+                                            header_.section_list_offset + sizeof(uint16_t)));
+
+  if (bytes_read < number_of_sections * sizeof(CaptureFileSection)) {
+    return ErrorMessage{absl::StrFormat(
+        "Unexpected EOF while reading section list: section list size=%d, bytes available=%d",
+        number_of_sections * sizeof(CaptureFileSection), bytes_read)};
+  }
+
+  section_list_ = std::move(section_list);
+
+  return outcome::success();
+}
+
+const std::vector<CaptureFileSection>& CaptureFileImpl::GetSectionList() const {
```

Comment:
Personal preference (feel free to ignore): I would implement the getter already at the definition above.